crates/core/src/logstore/factories.rs

```diff
@@ -25,13 +26,22 @@ pub trait ObjectStoreFactory: Send + Sync {
     /// corresponding to the path segment of the URL.
     ///
     /// The store should __NOT__ apply the decorations via the passed `options`
+    #[cfg(feature = "cloud")]
     fn parse_url_opts(
         &self,
         url: &Url,
         options: &HashMap<String, String>,
         retry: &RetryConfig,
         handle: Option<Handle>,
     ) -> DeltaResult<(ObjectStoreRef, Path)>;
+
+    #[cfg(not(feature = "cloud"))]
+    fn parse_url_opts(
```

Comment:
Yeah, this is the part that causes trouble. The trait having multiple implementations of the same method gets hairy.

Once you're done tinkering with this PR, I'll extend it to put this behind a field variant to avoid issues in the future

Reply:
How about something like:

`#[cfg(feature = "cloud")]
pub struct RetryCapsule(RetryConfig);
impl Deref for RetryCapsule {
    type Target = RetryConfig;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}
#[cfg(not(feature = "cloud"))]
pub struct RetryCapsule();`

And then:

`pub trait ObjectStoreFactory: Send + Sync {
    fn parse_url_opts(
        &self,
        url: &Url,
        options: &HashMap<String, String>,
        retry: &RetryCapsule,
        handle: Option<Handle>,
    ) -> DeltaResult<(ObjectStoreRef, Path)>;
}`

Reply:
> What I am thinking is putting a more generic/extendable enum in place for passing other configurations through in the future too

Got it.
So how do we handle this, should I complete this PR now and you'll take over for the extendable implementation?